src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs

```diff
@@ -85,9 +85,38 @@ internal Task ConnectAsync(EndPoint remoteEP)
             return saea.ConnectAsync(this).AsTask();
         }
 
+        internal async ValueTask ConnectAsync(EndPoint remoteEP, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Use _singleBufferReceiveEventArgs so the AwaitableSocketAsyncEventArgs can be re-used later for receives.
+            AwaitableSocketAsyncEventArgs saea =
+                Interlocked.Exchange(ref _singleBufferReceiveEventArgs, null) ??
+                new AwaitableSocketAsyncEventArgs(this, isReceiveForCaching: true);
+
+            saea.RemoteEndPoint = remoteEP;
+
+            try
+            {
+                using (cancellationToken.UnsafeRegister(o => CancelConnectAsync((SocketAsyncEventArgs)o!), saea))
+                {
+                    await saea.ConnectAsync(this).ConfigureAwait(false);
+                }
```

Comment:
If there's any reasonable chance ConnectAsync may complete synchronously (or even just really fast), it'd be slightly cheaper to do:
```suggestion
                ValueTask vt = saea.ConnectAsync(this);
                if (vt.IsCompleted)
                {
                    vt.GetAwaiter().GetResult();
                }
                else
                {
                    using (cancellationToken.UnsafeRegister(o => CancelConnectAsync((SocketAsyncEventArgs)o!), saea))
                    {
                        await vt;
                    }
                }
```
so that we avoid the UnsafeRegister if it completes synchronously.